third-party/spamassassin/src/main/java/org/apache/james/spamassassin/SpamAssassinInvoker.java

```diff
@@ -111,48 +114,69 @@ public SpamAssassinInvoker(MetricFactory metricFactory, String spamdHost, int sp
     }
 
     private Mono<SpamAssassinResult> scanMailWithAdditionalHeaders(MimeMessage message, String... additionalHeaders) {
-       return Mono.fromCallable(() -> {
-            try (Socket socket = new Socket(spamdHost, spamdPort);
-                 OutputStream out = socket.getOutputStream();
-                 BufferedOutputStream bufferedOutputStream = new BufferedOutputStream(out);
-                 PrintWriter writer = new PrintWriter(bufferedOutputStream);
-                 BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {
-
-                LOGGER.debug("Sending email {} for spam check", message.getMessageID());
-
-                writer.write("CHECK SPAMC/1.2");
-                writer.write(CRLF);
-
-                Arrays.stream(additionalHeaders)
-                    .forEach(header -> {
-                        writer.write(header);
-                        writer.write(CRLF);
-                    });
-
-                writer.write(CRLF);
-                writer.flush();
-
-                // pass the message to spamd
-                message.writeTo(out);
-                out.flush();
-                socket.shutdownOutput();
-
-                SpamAssassinResult spamAssassinResult = in.lines()
-                    .filter(this::isSpam)
-                    .map(this::processSpam)
-                    .findFirst()
-                    .orElse(SpamAssassinResult.empty());
-
-                LOGGER.debug("spam check result: {}", spamAssassinResult);
-                return spamAssassinResult;
-            } catch (UnknownHostException e) {
-                throw new MessagingException("Error communicating with spamd. Unknown host: " + spamdHost);
-            } catch (IOException | MessagingException e) {
-                throw new MessagingException("Error communicating with spamd on " + spamdHost + ":" + spamdPort, e);
-            }
+        Mono<? extends Connection> connectionMono = createConnection();
+
+        return connectionMono.flatMap(connection -> {
+            sendScanMailRequest(message, connection, additionalHeaders);
+            Mono<SpamAssassinResult> result = getSpamAssassinResultMono(connection.inbound());
+            return result;
         });
     }
 
+    private Mono<? extends Connection> createConnection() {
+        return TcpClient.create()
+            .host(spamdHost)
+            .port(spamdPort)
+            .doOnConnected(c -> c.addHandlerLast("codec",
+                new LineBasedFrameDecoder(8 * 1024)))
```

Comment:
can you explain me why we need this hardcoded value and what will happen if a given incoming line is longer than that?

Reply:
Reactor netty need a max length for the line decoder. It throw an exception if the line is too long.
8KB is a safe value since the spam assassin response barely do more than 30 chars.

Reply:
I extracted a constant for that